# Capítulo 5

## Repetições

### Exercício 5.1

Modifique o programa (exemplo contido no livro) para imprimir os números de 1 a 100.

In [3]:
i = 1

while i <= 100:
    print(i, end=', ')
    i += 1
    if i == 100:
        print(100)
        break

1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100


### Exercício 5.2

Modifique o programa para exibir os números de 50 a 100.

In [1]:
i = 50

while i <= 100:
    print(i, end=', ')
    i += 1
    if i == 100:
        print(100)
        break

50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100


### Exercício 5.3

Faça um programa para escrever a contagem regressiva do lançamento de um foguete. O programa deve imprimir 10, 9, 8 ... 1, 0 e Fogo! na tela.

In [2]:
i = 10

while i >= 0:
    print(i, end=', ')
    i-= 1
    if i == 0:
        print(0, "e Fogo!")
        break

10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 e Fogo!


### Exercício 5.4

Modifique o programa anterior para imprimir de 1 até o número digitado pelo usuário, mas, dessa vez, apenas os números ímpares.

In [ ]:
i = 1
n = int(input("Digite o número final: "))

while i <= n:
    if i % 2 != 0:
        if i == n:
            print(n)
            break
        print(i, end=', ')
    i += 1

### Exercício 5.5

Reescreva o programa anterior para escrever os 10 primeiros múltiplos de 3.

In [3]:
i = 1
n = 3
while i < 10:
    print(n, end=', ')
    n += 3
    i += 1
    if i == 10:
        print(n)
        break

3, 6, 9, 12, 15, 18, 21, 24, 27, 30


### Exercício 5.6

Altere o programa anterior (livro) para exibir os resultados no mesmo formato de uma tabuada: 2x1 = 2, 2x2 = 4...

In [4]:
i = 1

while i <= 10:
    print("2 x {} = {}".format(i, 2 * i))
    i += 1

2 x 1 = 2
2 x 2 = 4
2 x 3 = 6
2 x 4 = 8
2 x 5 = 10
2 x 6 = 12
2 x 7 = 14
2 x 8 = 16
2 x 9 = 18
2 x 10 = 20


### Exercício 5.7

Modifique o programa anterior para que o usuário digite o início e o fim da tabuada.

In [5]:
i = int(input("Digite o início da tabuada: "))
j = int(input("Digite o fim da tabuada: "))

while i <= j:
    print("2 x {} = {}".format(i, 2 * i))
    i += 1

Digite o início da tabuada: 5
Digite o fim da tabuada: 9
2 x 5 = 10
2 x 6 = 12
2 x 7 = 14
2 x 8 = 16
2 x 9 = 18


### Exercício 5.8

Escreva um programa que leia dois números. Imprima o resultado da multiplicação do primeiro pelo segundo. Utilize apenas os operadores de soma e subtração para calcular o resultado. Lembre-se de que podemos entender a multiplicação de dois números como somas sucessivas de um deles. Assim, 4 x 5 = 5 + 5 + 5 + 5 +5 = 4 + 4 + 4 + 4 +4

In [6]:
a = int(input("Digite o primeiro valor: "))
b = int(input("Digite o segundo valor: "))
count = 1
aux = a

while count < b:
    a += aux
    count += 1
print("Resultado: {}".format(a))

Digite o primeiro valor: 9
Digite o segundo valor: 5
Resultado: 45


### Exercício 5.9

Escreva um programa que leia dois números. Imprima a divisão inteira do primeiro pelo segundo, assim como o resto da divisão. Utilize apenas os operadores de soma e subtração para calcular o resultado. Lembre-se de que podemos entender o quociente da divisão de dois números como a quantidade de vezes que podemos retirar o divisor  do dividendo. Logo, 20 ÷ 4 = 5, uma vez que podemos subtrair 4 cinco vezes de 20.

In [7]:
a = int(input("Digite o primeiro valor: "))
b = int(input("Digite o segundo valor: "))
cont = 0
while a >= b:
    a -= b
    cont += 1
print("O resultado da divisão é {}, sendo o seu resto o valor {}".format(cont, a))

Digite o primeiro valor: 56
Digite o segundo valor: 48
O resultado da divisão é 1, sendo o seu resto o valor 8


### Exercício 5.10

Modifique o programa da listagem 5.10 (livro) para que aceite restas com letras maiúsculas e minúsculas em todas as questões.

In [8]:
pontos = 0
questao = 1

while questao <= 3:
    resposta = input("Digite a resposta da questão {} (A,B,C ou D): ".format(questao))
    if questao == 1 and resposta == "b" or resposta == "B":
        pontos += 1

    if questao == 2 and resposta == "a" or resposta == "A":
        pontos += 1

    if questao == 3 and resposta == "d" or resposta == "D":
        pontos += 1
    
    questao += 1
    
print("Você fez {} pontos".format(pontos))

Digite a resposta da questão 1 (A,B,C ou D): a
Digite a resposta da questão 2 (A,B,C ou D): b
Digite a resposta da questão 3 (A,B,C ou D): a
Você fez 0 pontos


### Exercício 5.11

Escreva um programa que pergunte o despósito inicial e a taxa de juros de uma poupança. Exiba os valores mês para os 24 primeiros meses. Escreva o total ganho com juros no período.

In [9]:
deposito_inicial = float(input("Digite o depósito inicial: "))
taxa_jurso = float(input("Digite a taxa de júros mensal: "))
i = 0
total = 0
rendimento = 0
while i < 24:
    rendimento = deposito_inicial * taxa_jurso / 100
    deposito_inicial += rendimento
    i += 1
    total += rendimento
    print("O rendimento no mês {} foi de: {:.2f}".format(i, rendimento))
print("O rendimento total da aplicação foi de:  {:.2f}".format(total))


Digite o depósito inicial: 1000
Digite a taxa de júros mensal: 0.5
O rendimento no mês 1 foi de: 5.00
O rendimento no mês 2 foi de: 5.03
O rendimento no mês 3 foi de: 5.05
O rendimento no mês 4 foi de: 5.08
O rendimento no mês 5 foi de: 5.10
O rendimento no mês 6 foi de: 5.13
O rendimento no mês 7 foi de: 5.15
O rendimento no mês 8 foi de: 5.18
O rendimento no mês 9 foi de: 5.20
O rendimento no mês 10 foi de: 5.23
O rendimento no mês 11 foi de: 5.26
O rendimento no mês 12 foi de: 5.28
O rendimento no mês 13 foi de: 5.31
O rendimento no mês 14 foi de: 5.33
O rendimento no mês 15 foi de: 5.36
O rendimento no mês 16 foi de: 5.39
O rendimento no mês 17 foi de: 5.42
O rendimento no mês 18 foi de: 5.44
O rendimento no mês 19 foi de: 5.47
O rendimento no mês 20 foi de: 5.50
O rendimento no mês 21 foi de: 5.52
O rendimento no mês 22 foi de: 5.55
O rendimento no mês 23 foi de: 5.58
O rendimento no mês 24 foi de: 5.61
O rendimento total da aplicação foi de:  127.16


### Exercício 5.12

Altere o programa anterior de forma a perguntar também o valor depositado mensalmente. Esse valor será depositado no início de cada mês, e você deve conderá-lo para o cálculo de juros.

In [1]:
deposito_inicial = float(input("Digite o depósito inicial: "))
taxa_juros = float(input("Digite a taxa de júros mensal: "))
i = 0
total = 0
rendimento = 0
deposito_mensal = float(input("Digite o valor a ser depositado mensalmente: "))
while i < 24:
    rendimento = deposito_inicial * taxa_juros / 100
    deposito_inicial += rendimento + deposito_mensal
    i += 1
    total += rendimento
    print("O rendimento no mês {} foi de: {:.2f}".format(i, rendimento))
print("O rendimento total da aplicação foi de:  {:.2f}".format(total))

Digite o depósito inicial: 1000
Digite a taxa de júros mensal: 10
Digite o valor a ser depositado mensalmente: 120
O rendimento no mês 1 foi de: 100.00
O rendimento no mês 2 foi de: 122.00
O rendimento no mês 3 foi de: 146.20
O rendimento no mês 4 foi de: 172.82
O rendimento no mês 5 foi de: 202.10
O rendimento no mês 6 foi de: 234.31
O rendimento no mês 7 foi de: 269.74
O rendimento no mês 8 foi de: 308.72
O rendimento no mês 9 foi de: 351.59
O rendimento no mês 10 foi de: 398.75
O rendimento no mês 11 foi de: 450.62
O rendimento no mês 12 foi de: 507.69
O rendimento no mês 13 foi de: 570.45
O rendimento no mês 14 foi de: 639.50
O rendimento no mês 15 foi de: 715.45
O rendimento no mês 16 foi de: 798.99
O rendimento no mês 17 foi de: 890.89
O rendimento no mês 18 foi de: 991.98
O rendimento no mês 19 foi de: 1103.18
O rendimento no mês 20 foi de: 1225.50
O rendimento no mês 21 foi de: 1360.05
O rendimento no mês 22 foi de: 1508.05
O rendimento no mês 23 foi de: 1670.86
O rendimento no

### Exercício 5.13

Escreva um programa que pergunte o valor inicial de uma dívida e o juro mensal. Pergunte também o valor mensal que será pago. Imprima o número de meses para que a dívida seja paga, o total pago e o total de juros pago.

In [1]:
divida_inicial = float(input("Digite o valor da dívida: "))
juro_mensal = float(input("Digite o valor do juro mensal: "))
pagamento_mensal = float(
    input("Digite o valor que será pago mensalmente para quitar a dívida: "))
meses = 0
juros_pagos = 0

if pagamento_mensal <= divida_inicial * juro_mensal / 100:
    print("O valor pago mensalmente não será suficiente"
          "para pagar a sua dívida.")
else:
    while divida_inicial > pagamento_mensal:
        juros = divida_inicial * juro_mensal / 100
        divida_inicial += juros
        divida_inicial -= pagamento_mensal
        meses += 1
        juros_pagos += juros

print("Total de juros pagos durante o período {:.2f}".format(juros_pagos))
if divida_inicial > 0:
    print("Total de mêses para que a parcela seja paga: {}. Mais 1 (uma) "
          "parcela com o valor remanecente no valor de {:.2f}".format(meses, divida_inicial))
else:
    print("Total de mêses para que a parcela seja paga {}".format(meses))
print("Total pago {:.2f}".format(
    (meses * pagamento_mensal + divida_inicial + juros_pagos)))


Digite o valor da dívida: 1000
Digite o valor do juro mensal: 5
Digite o valor que será pago mensalmente para quitar a dívida: 200
Total de juros pagos durante o período 171.16
Total de mêses para que a parcela seja paga: 5. Mais 1 (uma) parcela com o valor remanecente no valor de 171.16
Total pago 1342.31


### Exercício 5.14

Escreva um programa que leia números inteiros do teclado. O programa deve ler os números até que o usuário digite 0. No final da execução, exiba a quantidade de números digitados, assim como a sua soma e a média aritmética.

In [4]:
soma = 0
i = 0

while True:
    numero = int(input("Digite o número: "))
    if numero != 0:
        soma += numero
        i += 1
    else:
        break
print("Números digitados: {}\nSoma: {}\nMédia aritmética: {}".format(
    i, soma, (soma / i)))

Digite o número: 8
Digite o número: 8
Digite o número: 0
Números digitados: 2
Soma: 16
Média aritmética: 8.0


### Exercício 5.15

Escreva um programa para controlar uma pequena máquina registradora. Você deve solicitar ao usuário que digite o código do produto e a quantidade comprada. Utilize a tabela de código abaixo para obter o preço de cada produto.

|Código|  Preço |
|---|---|
| 1  | 0.50  |
| 2  | 1.00  |
| 3  | 4.00  |
| 5  | 7.00  |
| 9  | 8.00  |

Seu programa deve exibir o total das compras depois que o usuário digitar 0. Qualquer outro código deve gerar uma mensagem de erro "Código inválido"

In [5]:
codigo_produto = 0
quantidade = 0
total = 0

while True:
    while True:
        codigo_produto = int(input("Informe o código do produto: "))
        if codigo_produto < 0 or codigo_produto == 4 or codigo_produto > 5 and codigo_produto < 9 or codigo_produto >= 10:
            print("O código do produto é inválido")
        else:
            break

    if codigo_produto == 0:
        break

    quantidade = float(input("Quantidade comprada: "))
    
    if codigo_produto == 1:
        total += quantidade * 0.50
    elif codigo_produto == 2:
        total += quantidade * 1.00
    elif codigo_produto == 3:
        total += quantidade * 4.00
    elif codigo_produto == 5:
        total += quantidade * 7.00
    elif codigo_produto == 9:
        total += quantidade * 8.00

print("O total da compra foi: R${:.2f}".format(total))

Informe o código do produto: 1
Quantidade comprada: 2
Informe o código do produto: 3
Quantidade comprada: 1
Informe o código do produto: 9
Quantidade comprada: 0
Informe o código do produto: 0
O total da compra foi: R$5.00


### Exercício 5.16

Execute o programa da listagem 5.14 (livro) para os seguintes valores: 501, 745, 384, 2, 7 e 1

**Resposta**: O programa funciona corretamente

### Exercício 5.17

O que acontece se digitarmos 0 no valor a pagar?

***Resposta***: O programa imprime um valor incorreto.

### Exercício 5.18

Modifique o programa (listagem no livro) para que trabalhe também com notas de 100 reais.

In [1]:
valor = int(input("Digite o valor a pagar: "))
cedulas = 0
atual = 100
apagar = valor

while True:
    if atual <= apagar:
        apagar -= atual
        cedulas += 1
    else:
        print("{} cédulas de {}".format(cedulas, atual))

        if apagar == 0:
            break
        if atual == 100:
            atual = 50
        elif atual == 50:
            atual = 20
        elif atual == 20:
            atual = 10
        elif atual == 10:
            atual = 5
        elif atual == 5:
            atual = 1
        cedulas = 0
    

Digite o valor a pagar: 153
1 cédulas de 100
1 cédulas de 50
0 cédulas de 20
0 cédulas de 10
0 cédulas de 5
3 cédulas de 1


### Exercício 5.19

Modifique o programa para que também trabalhe com valores de moedas decimais.

In [3]:
valor = float(input("Digite o valor: "))
cedulas = 0
atual = 50
apagar = valor

while True:
    if atual <= apagar:
        apagar -= atual
        cedulas += 1
    else:
        if atual >= 1:
            print("{} cédulas de {:.2f}".format(cedulas, atual))
        else:
            if atual == 0.01:
                print("{} moedas de {:.2f}".format((cedulas + 1), atual))
                break
            else:
                print("{} moedas de {:.2f}".format(cedulas, atual))

        if apagar == 0:
            break
        if atual == 50:
            atual = 20
        elif atual == 20:
            atual = 10
        elif atual == 10:
            atual = 5
        elif atual == 5:
            atual = 1
        elif atual == 1:
            atual = 0.50
        elif atual == 0.50:
            atual = 0.25
        elif atual == 0.25:
            atual = 0.10
        elif atual == 0.10:
            atual = 0.05
        elif atual == 0.05:
            atual = 0.01

        cedulas = 0

Digite o valor: 59.73
1 cédulas de 50.00
0 cédulas de 20.00
0 cédulas de 10.00
1 cédulas de 5.00
4 cédulas de 1.00
1 moedas de 0.50
0 moedas de 0.25
2 moedas de 0.10
0 moedas de 0.05
3 moedas de 0.01


### Exercício 5.20

O que acontece se digitarmos 0.001 no programa anterior? Caso não funcione, altere-o de para corrigir o problema.

**Resposta**: O programa funciona com 1 centavo.

### Exercício 5.21

Reescreva o programa para que ele seja reexecutado até o que usuário digite 0.

In [4]:
while True:
    valor = float(input("Digite o valor: "))
    cedulas = 0
    atual = 50
    apagar = valor
    if valor == 0:
        break
    while True:
        if atual <= apagar:
            apagar -= atual
            cedulas += 1
        else:
            if atual >= 1:
                print("{} cédulas de {:.2f}".format(cedulas, atual))
            else:
                if atual == 0.01:
                    print("{} moedas de {:.2f}".format((cedulas + 1), atual))
                    break
                else:
                    print("{} moedas de {:.2f}".format(cedulas, atual))

            if apagar == 0:
                break
            if atual == 50:
                atual = 20
            elif atual == 20:
                atual = 10
            elif atual == 10:
                atual = 5
            elif atual == 5:
                atual = 1
            elif atual == 1:
                atual = 0.50
            elif atual == 0.50:
                atual = 0.25
            elif atual == 0.25:
                atual = 0.10
            elif atual == 0.10:
                atual = 0.05
            elif atual == 0.05:
                atual = 0.01

            cedulas = 0

Digite o valor: 2
0 cédulas de 50.00
0 cédulas de 20.00
0 cédulas de 10.00
0 cédulas de 5.00
2 cédulas de 1.00
Digite o valor: 0


### Exercício 5.22

Escreva um programa que exiba uma lista de opções: aritmética básica e **sair**. Imprima a tabuada da operação escolhida. Repita até que a opção saída seja escolhida.

In [5]:
while True:
    operacao = input("Escolha a operação (+ - ÷(/) x) ou digite 0 para sair: ")
    if operacao == "0":
        break
    elif operacao == "*":
        i = 1
        j = 1
        while i <= 10:
            i += 1
            while j <= 10:
                print("{} x {} = {}".format(j, i, (j * i)))
                j += 1
    elif operacao == "/":
        i = 1
        j = 1
        while i <= 10:
            i += 1
            while j <= 10:
                print("{} / {} = {}".format(j, i, (j / i)))
                j += 1
    elif operacao == "-":
        i = 1
        j = 1
        while i <= 10:
            i += 1
            while j <= 10:
                print("{} - {} = {}".format(j, i, (j - i)))
                j += 1
    elif operacao == "+":
        i = 1
        j = 1
        while i <= 10:
            i += 1
            while j <= 10:
                print("{} + {} = {}".format(j, i, (j + i)))
                j += 1

Escolha a operação (+ - ÷(/) x) ou digite 0 para sair: +
1 + 2 = 3
2 + 2 = 4
3 + 2 = 5
4 + 2 = 6
5 + 2 = 7
6 + 2 = 8
7 + 2 = 9
8 + 2 = 10
9 + 2 = 11
10 + 2 = 12
Escolha a operação (+ - ÷(/) x) ou digite 0 para sair: *
1 x 2 = 2
2 x 2 = 4
3 x 2 = 6
4 x 2 = 8
5 x 2 = 10
6 x 2 = 12
7 x 2 = 14
8 x 2 = 16
9 x 2 = 18
10 x 2 = 20
Escolha a operação (+ - ÷(/) x) ou digite 0 para sair: 0


### Exercício 5.23

Escreva um programa para verificar se ele é ou não um número primo. Para isso veja se o resto da divisão dele por 2 é diferente de 0. Depois verifique todos os números ímpares até chegar no valor do número.

In [1]:
n = int(input("Digite um número: "))

verificador = 3

while verificador < n:
    if n % 2 == 0:
        print("O número {} não é primo")
        break
    elif n % verificador == 0:
        print("O número {} não é primo")
        break
    verificador += 2
else:
    print("O número é primo")

Digite um número: 853
O número é primo


### Exercício 5.24

Modifique o programa anterior de forma a ler um número **n**. Imprima os *n* primeiros números primos.

In [1]:
ultimo = int(input("Digite um número para ver os primos antecessores: "))
n = 3
while True:
    verificador = 3

    if n % 2 == 0:
        break
    else:
        while verificador < n:
            if n % verificador == 0:
                break
            verificador += 2
        else:
            print(n)
    n += 2
    if n > ultimo:
        break

Digite um número para ver os primos antecessores: 50
3
5
7
11
13
17
19
23
29
31
37
41
43
47


### Exercício 5.25

Escreva um programa que calcule a raiz quadrada de um número. Utilize o método de Newton para obter um resultado aproximado. Sendo N o número a obter a raiz quadrada, condere a base b = 2. Calcule P usanndo a formula p = (b+(n/b))/2. Agora, calcule o quadrado de P. A cada passo faça b = p e recalcule P usando a fóruma apresentada. Pare quando a diferença absoluta entre N e o quadrado de P for menor que 0,0001.

In [4]:
n = int(input("Digite o número: "))
b = 2

while (n - (b * b)) / 2 > 0.00001:
    p = (b+(n/b))/2
    b = p

print(b)

Digite o número: 9
3.25


### Exercício 5.26

Escreva um programa que calcule o resto da divisão inteira entre dois números. Utilize apenas as operações de soma e subtração.

In [3]:
n = int(input("Digite o primeiro número: "))
n2 = int(input("Digite o segundo número: "))
i = 0

# ordenando os números para evitar erro de negativos
if n2 > n:
    aux = n
    n = n2
    n2 = aux

while n > n2:
    n = n - n2
    i += 1
print("O resto da divisão é: ", n)

Digite o primeiro número: 57
Digite o segundo número: 8
O resto da divisão é:  1


### Exercício 5.27

Escreva um programa que verifique se um número é palíndromo. Um número é palíndromo se contina o mesmo caso seus dígitos sejam invertidos.

Exemplos: 454, 10501

In [2]:
# Para a verificação eu optei por utilizar uma string, já que eu consigo percorrer o número e fazer a validação.

palindromo = str(input("Digite o número para verificar se ele é um palíndromo: "))
i = len(palindromo) - 1
verificacao = ""
while i >= 0:
    verificacao += palindromo[i]
    i -= 1
if verificacao == palindromo:
    print("O número é um palíndromo.")
else:
    print("O número não é um palíndromo.")

Digite o número para verificar se ele é um palíndromo: 454
O número é um palíndromo.
